File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def BFS(graph):
    n, m = len(graph), len(graph[0])
    visited = [[False] * m for _ in range(n)]
    visited[0][0] = True
    direction = [(0, 1), (0, -1), (1, 0), (-1, 0)]
    queue = deque([(0, 0, 1)])
    
    while queue:
        x, y, cnt = queue.popleft()
        if x == n - 1 and y == m - 1:
            return cnt
        for dx, dy in direction:
            nx, ny = x + dx, y + dy
            if 0 <= nx < n and 0 <= ny < m and graph[nx][ny] == 1 and not visited[nx][ny]:
                visited[nx][ny] = True
                queue.append((nx, ny, cnt + 1))
    return -1

def solution(maps):
    return BFS(maps)

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    private int bfs(int[][] maps, boolean[][] visited, int[] start){
        int rows = maps.length, cols = maps[0].length;
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        Queue<int[]> queue = new LinkedList<>();
        queue.add(start);
        while(!queue.isEmpty()){
            int[] current = queue.poll();
            int x = current[0], y = current[1], cost = current[2];
            if(x == rows-1 && y == cols -1) return cost;
            for(int[] dir : directions){
                int nx = x + dir[0], ny = y + dir[1];
                if(nx >= 0 && nx < rows && ny>=0 && ny < cols && !visited[nx][ny] && maps[nx][ny] != 0){
                    visited[nx][ny] = true;
                    queue.add(new int[]{nx, ny, cost+1});
                }
            }
        }
        return -1;
    }
    public int solution(int[][] maps) {
        boolean[][] visited = new boolean[maps.length][maps[0].length];
        return bfs(maps, visited, new int[]{0,0,1});
    }
}